### DISCUSSION TITLE
USB emulation

### DISCUSSION THREAD

**Discussion by Torinde** [86Box/86Box#3633] [General]
Recently there were a [lot of USB PRs](https://github.com/86Box/86Box/pulls?q=is%3Apr+is%3Aclosed+USB), but then it seems they [got reverted](https://github.com/86Box/86Box/commit/29c153d28743af963219252ae73fe671cfecf4f5) and the USB mouse emulation was rejected #3344

What's the story behind all that? What is the current status and what is planed?

Related discussion, but about pass-through is #2907

**Comment by iamgreaser** [comment]
The implementation was apparently several months worth of "grab code from Qemu then proceed to kludge it until it appears to vaguely resemble something that might work".

USB is a complex beast and if one takes that approach then it is guaranteed to fail miserably.

My advice for anyone who wants to try doing this properly is to read the crap out of the UHCI datasheet and/or OHCI datasheet, actually get to a point where the relevant parts of the USB specification start to make sense, and then do your own implementation.

I'm not going to attempt such a thing until I can get a UHCI driver working properly on my DOS rig, because going through all that would actually give me *some* understanding of what's going on.

**Reply by Torinde** [reply]
DOS USB drivers:
- [USBDDOS](https://github.com/crazii/USBDDOS)
- [USBDOS](https://bretjohnson.us)
- [DOSUSB](http://www.dosusb.net)
